[BUG] Fixed `AlignerDTW` when `window_size` needs to be passed given `window_type`

OmBiradar opened the pull request:
Fixes #8120 See also #8121

Fixed by correctly adjusting the `window_size` parameter handling in `AlignerDTW` and `AlignerDTWfromDist` classes.

The issue occurred when using `window_type='sakoechiba'` without properly passing the required `window_size` parameter to the underlying dtw-python package, resulting in `TypeError: sakoeChibaWindow() missing 1 required positional argument: 'window_size'`.

Changes:

- Created `window_size` parameter to pass parameters to `dtw()` function
- Added proper handling to include `window_size` parameter when needed
- Updated documentation to make `window_size` requirement clear
- Fixed examples to demonstrate correct usage with Sakoe-Chiba windowing

Discussion about the issue can be found [here](https://github.com/sktime/sktime/issues/8120#issuecomment-2773924362)

This ensures the advanced example in the documentation now works 

Please check if it works! I checked from my side and it works perfectely!


##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

fkiraly reviewed (changes requested):
Great fix, thanks!

Can I ask a few questions:

* are there any other window arguments that might be required? That is, do we at any point expect `window_args` to be non-empty but not with `window_size`?
* we should not add an argument in the middle since we do not want to break user code, see https://www.sktime.net/en/latest/developer_guide/deprecation.html
    * we could add new arguments at the end, and optionally carry out a change cycle to move it to the middle, see here: https://www.sktime.net/en/latest/developer_guide/deprecation.html#changing-the-sequence-of-parameters
    * one question to resolve first would be `window_arg` vs `window_ params`.

One change request: could we add the new parameter combination also to `get_fitted_params`, as a new parameter set, so we ensuer it is tested?

OmBiradar commented:
Thankyou for your detailed review @fkiraly

I will check into the actual logic as to what is the importance of `window_size`.
I think what I should do is refer the following:
- `dtw-python` library implementation
- previous implementation of `AlignerDTW` so as to understand the previously supported version

I think that the fix shouldn't affect the previous versions, it would be important to understand the exact expected output of the already in use function. 

I will see into what can be done for `get_fitted_params`, for now I think we should add the parameter list as it's a new parameter and also write some tests.

I think it might take some time to research and make valuable addition.

Additional node: Great documentation of `sktime` which even mentions how small changes should be taken care of without affecting the user base.

fkiraly commented:
> I will see into what can be done for get_fitted_params, for now I think we should add the parameter list as it's a new parameter and also write some tests.

Apologies, I meant `get_test_params` - so to add the new parameter set with `window_size` as a test parameter set.

OmBiradar commented:
Hey @fkiraly,

**A bit about what I found**:
- `window_size` is a must parameter for DTW algorithm.
- Generally people agree that `window_size` should be 10% of the size of the data.
- I checked previous versions of `sktime` and found the same bug over 

**What I conclude**:
- `window_size`, if not specified - should be set to have a default value of 10% of the dataset.
- Adding this to the `get_test_params` for testing might not be required if the default value of 10% is being followed, but anyhow - extra testing does no harm!
- This just might be a bug that existed from the previous versions and wasn't discovered.

**What I think should be done**:
- add the `window_size` parameter, but keep the default value to be 10% of the dataset size, thus it won't break any previous code either

**Ambitious thinking:**
(Just me thinking a bit too much about the topic)

Argument: _Why don't we just remove the `dtw-python` as a dependency and write the code in the repository itself._
Defense: The `dtw-python` library is being used here only for it's 1 function `dwt` while in fact it contains many functions like the ability to plot graphs etc. In short `dwt-python` is in itself a full utility library, thus just using it for a single function may seen redundant.
What can be done is maybe the `dwt-python` library could be made of use more by using it's other functions like graph creation, but then again I think it would differ from the style of graphs of `sktime`. 
Maybe we can add support for graphs etc in `sktime` 

Thinking: The distance matrix is updated after the `fit` function is run, thus to just get the distance matrix one is compeled to fit the model, which includes creation of the distance matrix along with additional computation which can reduce the efficiency of the code. Thus it might be better to have stand-alone functions `get_distance` and `get_alignment` work independent of the `fit` function, in the same time, the `fit` function can be modified to execute the `get_distance` and `get_alignment` functions one after the other.

fkiraly commented:
> might not be required if the default value of 10% is being followed, 

This is for coverage - if we add the parameter, we should add a parameter set that has it as non-default.

> add the window_size parameter, but keep the default value to be 10% of the dataset size, thus it won't break any previous code either

Ok, I see - maybe add it directly after `window_type` then, since the parameters after that cannot be used without a windowed approach, right?

> Why don't we just remove the dtw-python as a dependency and write the code in the repository itself.

It is one of the reference implementations of the DTW algorithm. Why reinvent the wheel.

> Thus it might be better to have stand-alone functions get_distance and get_alignment work independent of the fit function

This is separate from this PR - can you explain why that would be better?

OmBiradar commented:
> This is for coverage - if we add the parameter, we should add a parameter set that has it as non-default.

Sure, seems right!

> Ok, I see - maybe add it directly after window_type then, since the parameters after that cannot be used without a windowed approach, right?
Yes, I think the doc string should also be updated to mention the various values of other parameters which I found were missing.

> It is one of the reference implementations of the DTW algorithm. Why reinvent the wheel.
Yes your right, but it might raise errors if the `dtw-python` releases it's support later than `sktime` schedule.

> This is separate from this PR - can you explain why that would be better?
The DTW approach involves 2 steps - creating the distance matrix and then using the particular windowing method to calculate the appropriate alignment.

The `AlignerDTW` class first calculates the distance and alignment using the `dtw-python` library and then updates the distance matrix. Sometimes the user might just need the distance matrix and not alignment, maybe they need to align it after they calculate the distance.

I am unsure, this are just my thoughts.

fkiraly commented:
> The AlignerDTW class first calculates the distance and alignment using the dtw-python library and then updates the distance matrix.

I do not see the problem in terms of redundancy - to compute the alignment, you need to have computed the distance matrix. Whereas, if you have the distance matrix, you can get the alignment in linear time.

OmBiradar commented:
Hey @fkiraly 

I have updated the code to have 'window_size' have a default value of 10% of the series length.

Additionally I added the parameters set for Sochi-Chiba windowing in 'get_testing_params'.

I ran the tests and it was working!

Waiting for your say.

fkiraly commented:
I noticed some doctest examples were failing and fixed them (just a very minor issue)

fkiraly reviewed (changes requested):
The failure seems genuine - `window_size` should not be mutated by any method, the `self` attribute should always be identical to what is passed to `__init__`

OmBiradar commented:
@fkiraly just made it so that it doesn't change the actual `window_size` variable and instead the termprary variable... I think it should work this time

OmBiradar commented:
I think there is an error in the `ChronosForecaster` - about not loading the modle from Hugging face in the given timeout of 10s?

fkiraly reviewed (changes requested):
Great! Could you kindly include the `window_size` parameter in the docstring?

OmBiradar commented:
would this be fine? (first time writing doc strings.)

fkiraly reviewed (approved):
yes, great!
